test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java

```diff
@@ -0,0 +1,173 @@
+import java.security.cert.*;
+import java.security.cert.Certificate;
+
+public class ValidatePathWithURL {
+
+    private final X509Certificate rootCertificate;
+    private final X500Principal rootPrincipal;
+
+    public ValidatePathWithURL(String caAlias) throws Exception {
+        System.setProperty("com.sun.net.ssl.checkRevocation", "true");
+        Security.setProperty("ssl.TrustManagerFactory.algorithm", "SunPKIX");
+
+        // some test sites don't have correct hostname specified in test certificate
+        HttpsURLConnection.setDefaultHostnameVerifier(new CustomHostnameVerifier());
+
+        String FS = System.getProperty("file.separator");
+        String CACERTS_STORE =
+                System.getProperty("test.jdk") + FS + "lib" + FS + "security" + FS + "cacerts";
+
+        KeyStore cacerts = KeyStore.getInstance("PKCS12");
+        try (FileInputStream fis = new FileInputStream(CACERTS_STORE)) {
+            cacerts.load(fis, null);
+        }
+
+        rootCertificate = (X509Certificate) cacerts.getCertificate(caAlias);
+        rootPrincipal = rootCertificate.getSubjectX500Principal();
+    }
+
+    public static void enableOCSPOnly() {
+        System.setProperty("com.sun.security.enableCRLDP", "false");
+        Security.setProperty("ocsp.enable", "true");
+    }
+
+    public static void enableCRLOnly() {
+        System.setProperty("com.sun.security.enableCRLDP", "true");
+    }
+
+    public static void enableOCSPAndCRL() {
+        System.setProperty("com.sun.security.enableCRLDP", "true");
+        Security.setProperty("ocsp.enable", "true");
+    }
+
+    public void validateDomain(final String testURL,
+                               final boolean revokedCert)
+            throws CertificateException, NoSuchAlgorithmException, SignatureException,
+                    InvalidKeyException, NoSuchProviderException {
+        System.out.println();
+        System.out.println("===== Validate " + testURL + "=====");
+        if (!validateDomainCertChain(testURL, revokedCert)) {
+            throw new RuntimeException("Failed to validate " + testURL);
+        }
+        System.out.println("======> SUCCESS");
+    }
+
+    private boolean validateDomainCertChain(final String testURL,
+                                            final boolean revokedCert)
+            throws CertificateException, NoSuchAlgorithmException, SignatureException,
+                    InvalidKeyException, NoSuchProviderException {
+        HttpsURLConnection httpsURLConnection = null;
+        try {
+            URL url = new URL(testURL);
+            httpsURLConnection = (HttpsURLConnection) url.openConnection();
+            httpsURLConnection.setInstanceFollowRedirects(false);
+            httpsURLConnection.connect();
+
+            // certain that test certificate anchors to trusted CA for VALID certificate
+            // if the connection is successful
+            Certificate[] chain = httpsURLConnection.getServerCertificates();
```

Comment:
Mostly FYI - but another way to get the chain is to call the `getCertPath()` method of the thrown `CertPathValidatorException`. This doesn't include the root though.